# CNV-Finder

`GP2 â¤ï¸ Open Science ğŸ˜`

[![DOI](https://zenodo.org/badge/890546338.svg)](https://doi.org/10.5281/zenodo.14182563)

## Overview
CNV-Finder is a novel pipeline integrating a Long Short-Term Memory network on SNP array data to expedite large-scale identification of CNVs within predefined genomic regions. Manuscript coming soon!

## Installation
### Clone the repository:

````
git clone https://github.com/nvk23/CNV-Finder.git

cd CNV-Finder
````

### [Optional] Create a Conda environment:

````
conda create -n "cnv_finder" python=3.9.16 ipython

conda activate cnv_finder
````

### Install the required packages:

````
pip install -r requirements.txt
````

## Running the Pipeline
Open the `run_pipeline.ipynb` notebook and sequentially run through each cell to perform the 3 major processes in the pipeline: ML data preprocessing, application of pre-trained models/training new models on the prepared data, and the creation of app-ready files for visualizations in the CNV-Finder app.

### Available Parameters
For a more in-depth guide to the parameters available for each process, please read through the following documentation: `docs/parameter_guide.md`. 

## Project Structure
```
CNV_finder/
â”œâ”€â”€ app/
â”‚   â”œâ”€â”€ selections/
â”‚   â”œâ”€â”€ Home.py
â”‚   â””â”€â”€ variant_plots.py
â”‚
â”œâ”€â”€ modules/
â”‚   â”œâ”€â”€ cnv_finder/
â”‚   â”‚   â”œâ”€â”€ data_methods.py
â”‚   â”‚   â””â”€â”€ model_methods.py
â”‚   â”œâ”€â”€ run_data_prep.py
â”‚   â”œâ”€â”€ run_lstm_model.py
â”‚   â””â”€â”€ run_app_prep.py
â”‚
â”œâ”€â”€ ref_files/
â”‚   â”œâ”€â”€ models/
â”‚   â”‚   â”œâ”€â”€ final_del_5_50_combo4_lstm.keras
â”‚   â”‚   â”œâ”€â”€ final_dup_10_70_combo6_lstm.keras
â”‚   â”‚   â”œâ”€â”€ updated_del_5_50_combo4_lstm.keras
â”‚   â”‚   â”œâ”€â”€ updated_dup_10_70_combo6_lstm.keras
â”‚   â”‚   â”œâ”€â”€ prelim_del_5_50_combo4_lstm.keras
â”‚   â”‚   â””â”€â”€ prelim_dup_10_70_combo6_lstm.keras
â”‚   â”œâ”€â”€ custom_intervals.csv
â”‚   â”œâ”€â”€ glist_hg38_interval.csv
â”‚   â””â”€â”€ training_set_IDs.csv
â”‚
â”œâ”€â”€ example_data/
â”‚   â”œâ”€â”€ snp_metrics/
â”‚   â””â”€â”€ test_master_key.csv
â”‚
â”œâ”€â”€ testing/
â”‚   â”œâ”€â”€ app_ready/
â”‚   â”‚   â”œâ”€â”€ cohort/
â”‚   â”‚   â”‚   â”œâ”€â”€ final_del_model/
â”‚   â”‚   â”‚   â””â”€â”€ final_dup_model/
â”‚   â”œâ”€â”€ del/
â”‚   â”‚   â”œâ”€â”€ cohort/
â”‚   â”‚   â”‚   â””â”€â”€ gene/
â”‚   â””â”€â”€ dup/
â”‚       â””â”€â”€ cohort/
â”‚           â””â”€â”€ gene/
â”‚
â”œâ”€â”€ docs/
â”‚   â””â”€â”€ parameter_guide.md
â”‚
â”œâ”€â”€ run_pipeline.ipynb
â”œâ”€â”€ requirements.txt
â””â”€â”€ README.md
```

## Software
|               Software              |      Version(s)     |                       Resource URL                       |       RRID      |                                               Notes                                               |
|:-----------------------------------:|:-------------------:|:--------------------------------------------------------:|:---------------:|:-------------------------------------------------------------------------------------------------:|
|               Python Programming Language              |      3.9.16     |        http://www.python.org/        | RRID:SCR_008394 |                Refer to requirements.txt for necessary packages                |
